eth/catalyst/api.go

```diff
@@ -171,64 +173,83 @@ func (api *ConsensusAPI) makeEnv(parent *types.Block, header *types.Header) (*bl
 	return env, nil
 }
 
-func (api *ConsensusAPI) PreparePayload(params AssembleBlockParams) (*PayloadResponse, error) {
-	data, err := api.assembleBlock(params)
-	if err != nil {
-		return nil, err
-	}
-	id := len(api.preparedBlocks)
-	api.preparedBlocks[id] = data
-	return &PayloadResponse{PayloadID: uint64(id)}, nil
-}
-
-func (api *ConsensusAPI) GetPayload(PayloadID hexutil.Uint64) (*ExecutableData, error) {
-	data, ok := api.preparedBlocks[int(PayloadID)]
+func (api *ConsensusAPI) GetPayloadV1(PayloadID hexutil.Uint64) (*ExecutableDataV1, error) {
+	data, ok := api.preparedBlocks[uint64(PayloadID)]
 	if !ok {
 		return nil, &UnknownPayload
 	}
 	return data, nil
 }
 
-// ConsensusValidated is called to mark a block as valid, so
-// that data that is no longer needed can be removed.
-func (api *ConsensusAPI) ConsensusValidated(params ConsensusValidatedParams) error {
-	switch params.Status {
-	case VALID.Status:
-		return nil
-	case INVALID.Status:
-		// TODO (MariusVanDerWijden) delete the block from the bc
-		return nil
-	default:
-		return errors.New("invalid params.status")
+func (api *ConsensusAPI) ForkchoiceUpdatedV1(HeadBlockHash, SafeBlockHash, FinalizedBlockHash common.Hash, PayloadAttributes *PayloadAttributesV1) (GenericStringResponse, error) {
+	var emptyHash = common.Hash{}
+	if !bytes.Equal(HeadBlockHash[:], emptyHash[:]) {
+		if err := api.checkTerminalTotalDifficulty(HeadBlockHash); err != nil {
+			if block := api.eth.BlockChain().GetBlockByHash(HeadBlockHash); block == nil {
+				// TODO (MariusVanDerWijden) trigger sync
+				return SYNCING, nil
+			}
+			return INVALID, err
+		}
+		// If the finalized block is set, check if it is in our blockchain
+		if !bytes.Equal(FinalizedBlockHash[:], emptyHash[:]) {
+			if block := api.eth.BlockChain().GetBlockByHash(FinalizedBlockHash); block == nil {
+				// TODO (MariusVanDerWijden) trigger sync
+				return SYNCING, nil
+			}
+		}
+		// SetHead
+		if err := api.setHead(HeadBlockHash); err != nil {
+			return INVALID, err
+		}
+		// Assemble block (if needed)
+		if PayloadAttributes != nil {
+			data, err := api.assembleBlock(*PayloadAttributes)
+			if err != nil {
+				return INVALID, err
+			}
+			id := computePayloadId(HeadBlockHash, PayloadAttributes)
+			api.preparedBlocks[id] = data
+			log.Info("Created payload", "payloadid", id)
+			// TODO (MariusVanDerWijden) do something with the payloadID?
+			return VALID, nil
+		}
 	}
+	return VALID, nil
 }
 
-func (api *ConsensusAPI) ForkchoiceUpdated(params ForkChoiceParams) error {
-	var emptyHash = common.Hash{}
-	if !bytes.Equal(params.HeadBlockHash[:], emptyHash[:]) {
-		if err := api.checkTerminalTotalDifficulty(params.HeadBlockHash); err != nil {
-			return err
-		}
-		return api.setHead(params.HeadBlockHash)
+func computePayloadId(headBlockHash common.Hash, params *PayloadAttributesV1) uint64 {
```

Comment:
This changed in Kintsugi, to a 32-byte ID (just in case EL hashing is adopted by more CLs), and a byte prefixed to the input to version (to easily avoid collisions if we ever need to change the inputs).